examples/light_sleep.rs

```diff
@@ -0,0 +1,164 @@
+
+    let wakeup_reason = WakeupReason::get();
+    println!(
+        "wake up from light sleep due to {:?} which lasted for {:?}",
+        wakeup_reason,
+        time_after - time_before
+    );
+}
+
+fn main() -> anyhow::Result<()> {
+    esp_idf_sys::link_patches();
+
+    // run in a thread with increased stack size to prevent overflow
+    let builder = std::thread::Builder::new().stack_size(10 * 1024);
+    let th = builder.spawn(move || -> anyhow::Result<()> {
+        let peripherals = Peripherals::take().unwrap();
+
+        // RTC wakeup definitions
+        #[cfg(esp32)]
+        let rtc0 = PinDriver::rtc_input(peripherals.pins.gpio26)?;
+        #[cfg(esp32)]
+        let rtc1 = PinDriver::rtc_input(peripherals.pins.gpio27)?;
+
+        #[cfg(any(esp32s2, esp32s3))]
+        let rtc0 = PinDriver::rtc_input(peripherals.pins.gpio1)?;
+        #[cfg(any(esp32s2, esp32s3))]
+        let rtc1 = PinDriver::rtc_input(peripherals.pins.gpio2)?;
+
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        let rtc_pin0 = RtcWakeupPin { pindriver: &rtc0 };
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        let rtc_pin1 = RtcWakeupPin { pindriver: &rtc1 };
+        #[cfg(any(esp32, esp32s2, esp32s3))]
+        let rtc_wakeup = Some(RtcWakeup {
+            pins: EmptyRtcWakeupPins::chain(rtc_pin0).chain(rtc_pin1),
+            wake_level: RtcWakeLevel::AnyHigh,
+        });
+
+        // GPIO wakeup definitions
+        #[cfg(esp32)]
+        let gpio0 = PinDriver::input(peripherals.pins.gpio12)?;
+        #[cfg(esp32)]
+        let gpio1 = PinDriver::input(peripherals.pins.gpio14)?;
+
+        #[cfg(any(esp32s2, esp32s3))]
+        let gpio0 = PinDriver::input(peripherals.pins.gpio37)?;
+        #[cfg(any(esp32s2, esp32s3))]
+        let gpio1 = PinDriver::input(peripherals.pins.gpio38)?;
+
+        #[cfg(esp32c3)]
+        let gpio0 = PinDriver::input(peripherals.pins.gpio6)?;
+        #[cfg(esp32c3)]
+        let gpio1 = PinDriver::input(peripherals.pins.gpio7)?;
+
+        #[cfg(any(esp32, esp32c3, esp32s2, esp32s3))]
+        let gpio_pin0 = GpioWakeupPin {
+            pindriver: &gpio0,
+            wake_level: Level::High,
+        };
+        #[cfg(any(esp32, esp32c3, esp32s2, esp32s3))]
+        let gpio_pin1 = GpioWakeupPin {
+            pindriver: &gpio1,
+            wake_level: Level::High,
+        };
+        #[cfg(any(esp32, esp32c3, esp32s2, esp32s3))]
+        let gpio_wakeup = Some(GpioWakeup {
+            pins: EmptyGpioWakeupPins::chain(gpio_pin0).chain(gpio_pin1),
+        });
+        #[cfg(not(any(esp32, esp32c3, esp32s2, esp32s3)))]
+        let gpio_wakeup: Option<GpioWakeup> = None;
+
+        // UART definitions
+        let config = Config::new().baudrate(Hertz(115_200));
+        #[cfg(any(esp32))]
+        let uart = UartDriver::new(
+            peripherals.uart0,
+            peripherals.pins.gpio4,
+            peripherals.pins.gpio3,
+            Option::<gpio::Gpio0>::None,
```

Comment:
The type annotations is always a bit fishy :D Lets use a not so wired turbofish here like `None::<AnyIOPin>` should work, so its not a None of a specific pin.